src/www/app/outline_server.spec.ts

```diff
@@ -92,6 +92,216 @@ describe('OutlineServerRepository', () => {
+    expect(repo.getById('server-does-not-exist')).toBeUndefined();
+    expect(repo.getById('')).toBeUndefined();
+  });
+
+  it('renames servers', () => {
+    const NEW_SERVER_NAME = 'new server name';
+    const storage = new InMemoryStorage();
+    const repo = new OutlineServerRepository(getFakeServerFactory(), new EventQueue(), storage);
+    repo.add(shadowsocksConfigToAccessKey(CONFIG_0));
+    const server = repo.getAll()[0];
+    repo.rename(server.id, NEW_SERVER_NAME);
+    expect(server.name).toEqual(NEW_SERVER_NAME);
+    const serversStorage: ServersStorageV1 =
+        JSON.parse(storage.getItem(OutlineServerRepository.SERVERS_STORAGE_KEY));
+    const serverNames = serversStorage.map(s => s.name);
+    expect(serverNames).toContain(NEW_SERVER_NAME);
+  });
+
+  it('rename emits ServerRenamed event', async () => {
+    const NEW_SERVER_NAME = 'new server name';
+    const eventQueue = new EventQueue();
+    eventQueue.subscribe(ServerAdded, () => {});  // Silence dropped event warnings.
+    const repo =
+        new OutlineServerRepository(getFakeServerFactory(), eventQueue, new InMemoryStorage());
+    const accessKey = shadowsocksConfigToAccessKey(CONFIG_0);
+    repo.add(accessKey);
+    const server = repo.getAll()[0];
+    repo.rename(server.id, NEW_SERVER_NAME);
+
+    await new Promise<void>(resolve => {
+      eventQueue.subscribe(ServerRenamed, (event: ServerRenamed) => {
+        expect(event.server.name).toEqual(NEW_SERVER_NAME);
+        resolve();
+      });
+      eventQueue.startPublishing();
+    });
+  });
+
+  it('forgets servers', () => {
+    const storage = new InMemoryStorage();
+    const repo = new OutlineServerRepository(getFakeServerFactory(), new EventQueue(), storage);
+    repo.add(shadowsocksConfigToAccessKey(CONFIG_0));
+    repo.add(shadowsocksConfigToAccessKey(CONFIG_1));
+    const forgottenServerId = repo.getAll()[0].id;
+    repo.forget(forgottenServerId);
+    expect(repo.getById(forgottenServerId)).toBeUndefined();
+    const serverIds = repo.getAll().map(s => s.id);
+    expect(serverIds.length).toEqual(1);
+    expect(serverIds).not.toContain(forgottenServerId);
+    const serversStorage: ServersStorageV1 =
+        JSON.parse(storage.getItem(OutlineServerRepository.SERVERS_STORAGE_KEY));
+    const serverIdsStorage = serversStorage.map(s => s.id);
+    expect(serverIdsStorage).not.toContain(forgottenServerId);
+  });
+
+  it('forget emits ServerForgotten events', async () => {
+    const eventQueue = new EventQueue();
+    eventQueue.subscribe(ServerAdded, () => {});  // Silence dropped event warnings.
+    const repo =
+        new OutlineServerRepository(getFakeServerFactory(), eventQueue, new InMemoryStorage());
+    repo.add(shadowsocksConfigToAccessKey(CONFIG_0));
+    repo.add(shadowsocksConfigToAccessKey(CONFIG_1));
+    const forgottenServerId = repo.getAll()[0].id;
+    repo.forget(forgottenServerId);
+    await new Promise<void>(resolve => {
```

Comment:
Same as above for the other emit event tests.

Reply:
Done in 54492e2.